src/Chapter3.hs

```diff
@@ -1005,7 +1187,36 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
+newtype FKHealth = FKHealth Int deriving (Show)
+newtype FKAttack = FKAttack Int deriving (Show)
+newtype FKDefense = FKDefense Int deriving (Show)
+
+newtype FMHealth = FMHealth Int deriving (Show)
+newtype FMAttack = FMAttack Int deriving (Show)
+
+data FMonsterAction = MAttack | MRunAway deriving (Show)
+data FKnightAction = KAttack | KDrinkPotion | SpellCast deriving (Show)
+
+data FMonster = FMonster
+    {
+      monsterFHealth :: FMHealth,
+      monsterFAttack :: FMAttack,
+      monserActions :: [FMonsterAction]
+    } deriving (Show)
+  
+data FKnight = FKnight
+    {
+      knightFHealth :: FKHealth,
+      knightFAttack :: FKAttack,
+      knightFDefense :: FKDefense,
+      knightActions :: [FKnightAction]
+    } deriving (Show)
+
 
+{-
+I am not clear about happens when Attack happens, health boost happens, cast is spelt, etc. 
+Would like some help here!
+-}
```

Comment:
Sure, that's not a difficult problem! Your `KnightAction` data type looks like this:

```haskell
data FKnightAction = KAttack | KDrinkPotion | SpellCast
```

* On attack, you should attack another fighter using the value of the `knightAttack` field
* On drink potion, you should increase knights health
* On spell case, you should increase knight defence

I can see a problem with  potion drink or defence spell, because you don't know how much you should increase the corresponding values of the type. But this is solvable by storing these values inside a sum type:

```haskell
data FKnightAction
    = KAttack
    | KDrinkPotion Health
    | SpellCast Defence
```

Let me know if this helps you move forward of if you have any other questions :slightly_smiling_face: